### DISCUSSION TITLE
Data collections

### DISCUSSION THREAD

**Discussion by bholmesdev** [withastro/roadmap#477] [Proposal]
**Update:** this proposal has been moved to stage 2. [Continue discussion here](https://github.com/withastro/roadmap/issues/530)

## Summary
This proposal adds support for "data collections," or content collections using simple data storage formats (namely JSON and YAML). This also introduces a pluggable API to support any data format in the future.

## Background & Motivation
Content collections are restricted to supporting `.md` and `.mdx` files. This is limiting for other forms of data you may need to store, namely raw data formats like JSON, YAML, and TOML. Common use cases include storing blog or documentation authors separate from the content of each post.

The content collections API was built generically to support this future, choosing format-agnostic naming like `data` instead of `frontmatter` and `body` instead of `rawContent`. Because of this, expanding support to new data formats without API changes is possible to achieve.

## Goals
- Introduce JSON, YAML, and TOML collection support, configurable and queryable with similar APIs to content collections.
- Determine where data collections are stored. We may introduce a new `src/data/` directory distinct from `src/content/`, or simply allow data collections within `src/content/`.
- Introduce an integration API to 1) add new data formats / file extensions to content collections, and 2) implement how each format's collection entry info (data and body) is retrieved.

## Non-goals
- Pluggable APIs for content authoring formats like [Markdoc](https://markdoc.dev/) and [Org-mode](https://orgmode.org/). We recognize the need for support beyond Markdown and MDX, so this will be considered in a future proposal.
- A mechanism to "reference" collections from other collections. It's clear that some data collections are related to others. For example, you may want to connect an "author" property in a blog post to an "authors" collection entry. This will be explored in a future proposal.

### Using the built-in JSON format
Say you have a collection of blog post authors you would like to store as JSON. You can create a new collection under `src/content/` (or `src/data/`, pending conversation) like so:
```bash
src/content/
  authors/
    ben.json
    fred.json
    matthew.json
```

This collection can be configured with a schema like any other content collection:

```ts
// src/content/config.ts
import { defineCollection, z } from 'astro:content';

const authors = defineCollection({
	schema: z.object({
		name: z.string(),
		twitter: z.string().url(),
	})
});

export const collections = { authors };
```

It can also be queried like any other collection, this example using `getCollection('authors')`:

```tsx
---
import { getCollection } from 'astro:content';
const authors = await getCollection('authors');
---
<ul>
{authors.map(author => (
	<li>
		<a href={author.data.twitter}>{author.data.name}</a>
	</li>
)}
</ul>
```

Each entry will have the same type signature as a content collection entry, with the `slug` property and `render()` function removed:
```ts
type DataEntry = {
	id: 'file-1.json' | 'file-2.json'...
	collection: 'authors'
	data: ConfiguredSchema<'authors'>
	body: string;
}

type ContentEntry = {
	id: 'file-1.md' | 'file-2.jmd'...
	slug: 'file-1' | 'file-2'
	collection: 'blog'
	data: ConfiguredSchema<'blog'>
	body: string;
	render(): Promise<{ Content: AstroComponentFactory }>;
}
```

### Configuring and using a custom format
Say you want to store these authors as TOML instead. Assuming TOML were *not* built-in, you could configure the TOML collection type using the `addContentEntryType` integration hook:

```ts
// tomlIntegration.ts
import tomlParser from 'toml';

export const toml: AstroIntegration = () => ({
		'astro:config:setup'({ addContentEntryType }) {
			addContentEntryType({
				extensions: ['.toml'],
				getEntryInfo({ contents }) {
					const data = tomlParser.parse(contents);
					return {
						// The unparsed data object that can be passed to a Zod schema.
						data,
						// The body of the data file. You can pass the file contents unchanged, or strip unrelated metadata (ex. strip frontmatter from Markdown body)
						body: contents,
					}
				}
			});
		}
	}
});

// astro.config.mjs
import { toml } from './tomlIntegration'

export default {
	integrations: [toml()],
}
```

Now, you are free to use `.toml` files in your collections, even mixed alongside files of another type for smoother migration:

```bash
src/content/
	authors/
		ben.toml
		fred.toml
		matthew.json
```

**Comment by matthewp** [comment]
Love the direction here! Really like that it covers all of the use-cases. Couple of things:

- Not having `slug` and `render` feels a little off to me. Usually functions don't return different shapes like this. I would suggest:
  - Leave slug as it is. No, you don't need it for data, but it feels worse to special-case based on data type than it is to have an unhelpful property. In the future we can find a way to make it so that `slug` only exists on content that is intended to be used as a page (like, for example, a `definePage()` alternative to defineCollection).
  - Leave `render()` as a property but throw if it's called. Eventually we can allow integrations to define how render() works, but it's ok to punt on it and throw.
- It might be better if `getEntryInfo` is not an inline function like this, but instead is a module with an implementation. For example, `entryInfoEntrypoint: '@astrojs/yaml/entry'`.
  - This is the pattern we use for renderers and adapters, ie `clientEntrypoint`. The reason is so that it can work in SSR. I know that we do all of this during compilation and within Vite, but if we ever want to support remote content then this can't be done inline like the example.
     - this is more to ponder on, this could be a premature thing to do.

**Comment by FredKSchott** [comment]
Love this conceptually! A couple of comments:
- "YAML, and TOML" - I think it would be worth exploring the cost of out-of-the-box vs. opt-in for anything that isn't JSON here. YAML and TOML files seem to be more for config than data, so I actually would guess that these are pretty uncommon (not sure I've ever seen a directory of TOML before).
- What about single-file data collections? One use-case I'd love us to consider (either as an explicit goal or explicit out-of-scope) is a CSV file full of data. Is that supported? Is it not supported? I'd personally prefer to keep scope small and only support single-file data objects here in an MVP.
- Just throwing my hat in the ring for `src/data`. The main reason being that when the rules are different like this (ex: `render` and `slug` not working, `headings` is another one) it helps to acknowledge the difference vs. hiding it or pretending otherwise. I think these rules will only get larger over time (ex: interest in content that is page-level setting SEO metadata, content controlling its own slug/permalink, etc). I know you're intentionally avoiding taking a stance this early but just flagging my interest for when you do tackle this!

**Reply by bholmesdev** [reply]
Thanks for the feedback Fred! To the `src/data` point: I actually think we want `slug` and `headings` to be intentional opt-ins for `src/content` going forward. This is based on our earlier convos about `slug` and `id` serving competing needs, and `slug` assuming a use case (rendering routes) your content may not have. Rather than introducing a new folder convention, we could explore a content `type` field to extend capabilities:

```ts
// just data, will not be used as pages
const authors = defineCollection({
  schema: z.object({...}),
});

// blog collection, will be used as pages
const blog = defineCollection({
  type: 'page',  // permalinks unlocked
  permalink: ({ id }) => customSlugify(id),
  schema: z.object({ ... }),
});
```

There's [an active discussion on the `slug` config options here](https://github.com/withastro/roadmap/discussions/516#discussioncomment-5294535) where I provide the context from that earlier convo!

**Reply by matthewp** [reply]
> "YAML, and TOML" - I think it would be worth exploring the cost of out-of-the-box vs. opt-in for anything that isn't JSON here. YAML and TOML files seem to be more for config than data, so I actually would guess that these are pretty uncommon (not sure I've ever seen a directory of TOML before).

What is "data" exactly? Is an author's biography data? Is a step in a recipe data? If a human has to write something in a file, yaml and toml are more ergonomic than JSON. You can't write `"` in JSON without escaping. I use JSON for data that's been exported by some other system.
